src/sdk/pynni/tests/test_dependecy_aware.py

```diff
@@ -0,0 +1,148 @@
+
+
+class DependencyawareTest(TestCase):
+    @unittest.skipIf(torch.__version__ < "1.3.0", "not supported")
+    def test_dependency_aware_pruning(self):
+        model_zoo = ['resnet18']
+        pruners = [L1FilterPruner, L2FilterPruner, FPGMPruner, TaylorFOWeightFilterPruner,
+                   ActivationMeanRankFilterPruner, ActivationAPoZRankFilterPruner]
+        sparsity = 0.7
+        cfg_list = [{'op_types': ['Conv2d'], 'sparsity':sparsity}]
+        dummy_input = torch.ones(1, 3, 224, 224)
+        for model_name in model_zoo:
+            for pruner in pruners:
+                print('Testing on ', pruner)
+                ori_filters = {}
+                Model = getattr(models, model_name)
+                net = Model(pretrained=True, progress=False)
+                # record the number of the filter of each conv layer
+                for name, module in net.named_modules():
+                    if isinstance(module, nn.Conv2d):
+                        ori_filters[name] = module.out_channels
+
+                # for the pruners that based on the activations, we need feed
+                # enough data before we call the compress function.
+                optimizer = torch.optim.SGD(net.parameters(), lr=0.0001,
+                                 momentum=0.9,
+                                 weight_decay=4e-5)
+                criterion = torch.nn.CrossEntropyLoss()
+                tmp_pruner = pruner(
+                    net, cfg_list, optimizer, dependency_aware=True, dummy_input=dummy_input)
+                # train one single batch so that the the pruner can collect the
+                # statistic
+                optimizer.zero_grad()
+                out = net(dummy_input)
+                batchsize = dummy_input.size(0)
+                loss = criterion(out, torch.zeros(batchsize, dtype=torch.int64))
+                loss.backward()
+                optimizer.step()
+
+                tmp_pruner.compress()
+                tmp_pruner.export_model(MODEL_FILE, MASK_FILE)
+                # if we want to use the same model, we should unwrap the pruner before the speedup
+                tmp_pruner._unwrap_model()
+                ms = ModelSpeedup(net, dummy_input, MASK_FILE)
+                ms.speedup_model()
+                for name, module in net.named_modules():
+                    if isinstance(module, nn.Conv2d):
+                        expected = int(ori_filters[name] * (1-sparsity))
+                        filter_diff = abs(expected - module.out_channels)
+                        errmsg = '%s Ori: %d, Expected: %d, Real: %d' % (
+                            name, ori_filters[name], expected, module.out_channels)
+
+                        # because we are using the dependency-aware mode, so the number of the
+                        # filters after speedup should be ori_filters[name] * ( 1 - sparsity )
+                        print(errmsg)
+                        assert filter_diff <= 1, errmsg
```

Comment:
I tested it, this assertion failed for mobilenet_v2, would you please check it?

Reply:
Yes, it's normal that the filter_diff is larger than 1 in mobilenet_v2, because, in the depth-wise conv, there is a conv layer whose `group` == 'out_channels', which means there is only one filter in each `group`. Because of the group dependency, it is impossible for us to prune sparsity* ori_n_filter filters for this layer, and this is in line with expectations

Reply:
Got it!